File: src/modules/ai/lib/skills.ts
import { native, type DirEntry, type ReadResult } from "./native";
import { normalizeHandle, type Snippet } from "./snippets";

/**
 * Builtin skills — LangBot-style `skills/` directory convention (★ L4).
 *
 * A skill lives at `<workspaceRoot>/skills/<name>/skill.json` (or
 * `<name>.json`) with fields `{ name, description, prompt, handle?,
 * toolAllowlist? }`. Scanning is triggered at app boot (`useAiBootstrap`)
 * and by the settings page's "重新扫描" button; disabled builtins (tracked in
 * `snippetsStore.disabledBuiltinHandles`) are skipped so they don't revive.
 */

export type SkillFile = {
  name: string;
  description: string;
  prompt: string;
  handle?: string;
  toolAllowlist?: string[];
  /** S6: tools this skill requires to function (frontmatter `requires_tools`). */
  requiresTools?: string[];
  /** S6: env vars this skill requires (frontmatter `requires_env`). */
  requiresEnv?: string[];
  /** S6: when any of these tools is missing, skip injecting this skill. */
  fallbackForTools?: string[];
  /** P1-5 lifecycle metadata (curator). */
  agent_created?: boolean;
  created_at?: number;
  activity_ts?: number;
  usage_count?: number;
  /** Archived by the curator — the skill still exists but is hidden from the
   * active list (archive-only, never deleted). */
  archived?: boolean;
};

/** S6 activation state. */
export type SkillState = "ACTIVE" | "DEGRADED" | "UNAVAILABLE";

/** S6 prompt budget (PraisonAI SkillPromptBudget). */
export const SKILL_MAX_CHARS = 4096;
export const SKILL_MAX_COUNT = 50;

/**
 * Decide a skill's activation state given available tools/env (mirrors Rust
 * `skill_state`). A tool in `fallbackForTools` being missing → UNAVAILABLE;
 * any requirement missing → DEGRADED; else ACTIVE.
 */
export function skillState(
  skill: SkillFile,
  availableTools: string[],
  availableEnv: string[],
): SkillState {
  const hasTool = (t: string) => availableTools.includes(t);
  const hasEnv = (e: string) => availableEnv.includes(e);
  const missingTools = (skill.requiresTools ?? []).filter((t) => !hasTool(t));
  const missingEnv = (skill.requiresEnv ?? []).filter((e) => !hasEnv(e));
  const fallbackHit = (skill.fallbackForTools ?? []).some((t) => !hasTool(t));
  if (fallbackHit) return "UNAVAILABLE";
  if (missingTools.length > 0 || missingEnv.length > 0) return "DEGRADED";
  return "ACTIVE";
}

/** Truncate a skill body to the prompt budget with a marker. */
export function capSkillBody(body: string, maxChars = SKILL_MAX_CHARS): string {
  if (body.length <= maxChars) return body;
  return `${body.slice(0, maxChars)}…[truncated]`;
}

/** Serialize a snippet as a shareable skill.json payload (bundle export). */
export function serializeSkill(s: Snippet): string {
  const payload: SkillFile = {
    name: s.name,
    description: s.description,
    prompt: s.content,
    handle: s.handle || undefined,
    toolAllowlist: s.toolAllowlist,
  };
  return JSON.stringify(payload, null, 2);
}

/**
 * Import a skill.json payload into `<workspaceRoot>/skills/<name>.json`.
 * Validates first; refuses to clobber an existing file (mirrors theme-import
 * behavior — no silent overwrite).
 */
export async function importSkillToWorkspace(
  workspaceRoot: string | null,
  raw: string,
): Promise<{ ok: true; name: string } | { ok: false; error: string }> {
  const skill = parseSkillJson(raw);
  if (!skill) {
    return { ok: false, error: "invalid skill.json: name and prompt are required" };
  }
  if (!workspaceRoot) {
    return { ok: false, error: "no workspace root selected" };
  }
  const root = workspaceRoot.replace(/\/$/, "");
  const dir = `${root}/skills`;
  try {
    await native.createDir(dir);
  } catch {
    // Directory already exists — the common case.
  }
  const target = `${dir}/${skill.name}.json`;
  try {
    await native.readFile(target);
    return { ok: false, error: `skill "${skill.name}" already exists in skills/` };
  } catch {
    // Target does not exist; proceed.
  }
  try {
    await native.writeFile(target, JSON.stringify(skill, null, 2));
  } catch (e) {
    return { ok: false, error: String(e) };
  }
  return { ok: true, name: skill.name };
}

/**
 * Convert a SKILL.md (frontmatter + body) into a skill.json payload
 * (agentskills 兼容性评估 建议 3). Frontmatter supports: name, description,
 * handle, toolAllowlist (YAML-ish `name: value` lines or a comma/space list).
 * Returns null when the body is empty.
 */
export function convertSkillMd(md: string): SkillFile | null {
  const m = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/.exec(md);
  const front = m?.[1] ?? "";
  const body = (m?.[2] ?? md).trim();
  if (!body) return null;

  const fm: Record<string, string> = {};
  for (const line of front.split("\n")) {
    const i = line.indexOf(":");
    if (i <= 0) continue;
    const key = line.slice(0, i).trim().toLowerCase();
    const val = line.slice(i + 1).trim();
    fm[key] = val.replace(/^["']|["']$/g, "");
  }

  const name = (fm.name || fm.title || "").trim();
  const description = (fm.description || "").trim();
  const handle = (fm.handle || "").trim();
  const list = (k: string): string[] | undefined => {
    const raw = (fm[k] ?? "").trim();
    const items = raw.split(/[\s,]+/).map((s) => s.trim()).filter(Boolean);
    return items.length > 0 ? items : undefined;
  };

  return {
    name,
    description,
    prompt: body,
    handle: handle ? normalizeHandle(handle) : undefined,
    toolAllowlist: list("toolallowlist"),
    requiresTools: list("requirestools"),
    requiresEnv: list("requiresenv"),
    fallbackForTools: list("fallbackfortools"),
  };
}

/** Parse + validate a skill.json payload; returns null when malformed. */
export function parseSkillJson(raw: string): SkillFile | null {
  let v: unknown;
  try {
    v = JSON.parse(raw);
  } catch {
    return null;
  }
  if (!v || typeof v !== "object") return null;
  const o = v as Record<string, unknown>;
  const name = typeof o.name === "string" ? o.name.trim() : "";
  const prompt = typeof o.prompt === "string" ? o.prompt.trim() : "";
  if (!name || !prompt) return null;
  const strList = (k: string): string[] | undefined =>
    Array.isArray(o[k]) ? o[k].filter((x): x is string => typeof x === "string") : undefined;

  return {
    name,
    description: typeof o.description === "string" ? o.description : "",
    prompt,
    handle: typeof o.handle === "string" ? normalizeHandle(o.handle) : undefined,
    toolAllowlist: strList("toolAllowlist"),
    requiresTools: strList("requiresTools"),
    requiresEnv: strList("requiresEnv"),
    fallbackForTools: strList("fallbackForTools"),
    agent_created: o.agent_created === true,
    created_at: typeof o.created_at === "number" ? o.created_at : undefined,
    activity_ts: typeof o.activity_ts === "number" ? o.activity_ts : undefined,
    usage_count: typeof o.usage_count === "number" ? o.usage_count : 0,
    archived: o.archived === true,
  };
}

/**
 * Scan `<workspaceRoot>/skills/` for builtin skills and map them to
 * `builtin: true` Snippets. Absent directory / unreadable entries are
 * silently skipped (a workspace without `skills/` is the normal case).
 */
export async function scanSkillsDir(
  workspaceRoot: string | null,
): Promise<Snippet[]> {
  if (!workspaceRoot) return [];
  const root = workspaceRoot.replace(/\/$/, "");
  let entries: DirEntry[];
  try {
    entries = await native.readDir(`${root}/skills`);
  } catch {
    return [];
  }

  const out: Snippet[] = [];
  for (const entry of entries) {
    if (entry.kind === "dir") {
      const skill = await readSkillFile(`${root}/skills/${entry.name}/skill.json`);
      if (skill) out.push(skill);
    } else if (entry.kind === "file" && entry.name.endsWith(".json")) {
      const skill = await readSkillFile(`${root}/skills/${entry.name}`);
      if (skill) out.push(skill);
    }
  }
  return out;
}

async function readSkillFile(path: string): Promise<Snippet | null> {
  let result: ReadResult;
  try {
    result = await native.readFile(path);
  } catch {
    return null;
  }
  if (result.kind !== "text") return null;
  const skill = parseSkillJson(result.content);
  if (!skill) return null;
  // Archived skills still exist on disk but are hidden from the active set.
  if (skill.archived) return null;
  return {
    id: `builtin-${skill.name}`,
    handle: skill.handle ?? normalizeHandle(skill.name),
    name: skill.name,
    description: skill.description,
    content: skill.prompt,
    toolAllowlist: skill.toolAllowlist,
    builtin: true,
  };
}

File: src-tauri/src/modules/ai/skills.rs
//! Skill subsystem (P1): native skill.json parse/validate + safe name/path
//! handling for `create_skill`, mirroring the frontend `lib/skills.ts` and
//! `tools/createSkill.ts` contracts so existing test behavior is preserved.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillFile {
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub prompt: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub handle: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_allowlist: Option<Vec<String>>,
    /// S6: tools this skill requires to function (frontmatter `requires_tools`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires_tools: Option<Vec<String>>,
    /// S6: env vars this skill requires (frontmatter `requires_env`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires_env: Option<Vec<String>>,
    /// S6: `fallback_for_tools` — when one of these tools is missing, the
    /// skill should not be injected (its body presumes that tool).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fallback_for_tools: Option<Vec<String>>,
    #[serde(default)]
    pub agent_created: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub created_at: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub activity_ts: Option<u64>,
    #[serde(default)]
    pub usage_count: u64,
    #[serde(default)]
    pub archived: bool,
}

/// S6 activation state.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum SkillState {
    Active,
    Degraded,
    Unavailable,
}

/// S6 prompt budget (PraisonAI `SkillPromptBudget`).
pub const SKILL_MAX_CHARS: usize = 4096;
pub const SKILL_MAX_COUNT: usize = 50;

/// Decide a skill's activation state given which tools/env are available.
/// - All requirements present → Active.
/// - Some missing but no `fallback_for_tools` hit → Degraded (still inject).
/// - A tool listed in `fallback_for_tools` is missing → Unavailable (skip
///   injection — its body presumes that tool).
pub fn skill_state(skill: &SkillFile, available_tools: &[String], available_env: &[String]) -> SkillState {
    let has_tool = |t: &str| available_tools.iter().any(|a| a == t);
    let has_env = |e: &str| available_env.iter().any(|a| a == e);

    let missing_tools: Vec<&str> = skill
        .requires_tools
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .map(String::as_str)
        .filter(|t| !has_tool(t))
        .collect();
    let missing_env: Vec<&str> = skill
        .requires_env
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .map(String::as_str)
        .filter(|e| !has_env(e))
        .collect();

    let fallback_hit = skill
        .fallback_for_tools
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .any(|t| !has_tool(t));

    if fallback_hit {
        return SkillState::Unavailable;
    }
    if !missing_tools.is_empty() || !missing_env.is_empty() {
        return SkillState::Degraded;
    }
    SkillState::Active
}

/// Truncate a skill body to the prompt budget (chars), with a marker.
pub fn cap_skill_body(body: &str, max_chars: usize) -> String {
    if body.chars().count() <= max_chars {
        return body.to_string();
    }
    let head: String = body.chars().take(max_chars).collect();
    format!("{head}…[truncated]")
}

/// Parse + validate a skill.json payload; returns `None` when malformed
/// (missing/invalid name or prompt). Mirrors `parseSkillJson`.
pub fn parse_skill_json(raw: &str) -> Option<SkillFile> {
    let v: serde_json::Value = serde_json::from_str(raw).ok()?;
    let o = v.as_object()?;
    let name = o.get("name").and_then(serde_json::Value::as_str).unwrap_or("").trim().to_string();
    let prompt = o.get("prompt").and_then(serde_json::Value::as_str).unwrap_or("").trim().to_string();
    if name.is_empty() || prompt.is_empty() {
        return None;
    }
    let description = o.get("description").and_then(serde_json::Value::as_str).unwrap_or("").to_string();
    let handle = o.get("handle").and_then(serde_json::Value::as_str).map(normalize_handle);
    let tool_allowlist = o
        .get("toolAllowlist")
        .and_then(serde_json::Value::as_array)
        .map(|arr| {
            arr.iter()
                .filter_map(serde_json::Value::as_str)
                .map(str::to_string)
                .collect()
        });
    let string_list = |k: &str| {
        o.get(k)
            .and_then(serde_json::Value::as_array)
            .map(|arr| {
                arr.iter()
                    .filter_map(serde_json::Value::as_str)
                    .map(str::to_string)
                    .collect()
            })
    };
    Some(SkillFile {
        name,
        description,
        prompt,
        handle: handle.filter(|h| !h.is_empty()),
        tool_allowlist,
        requires_tools: string_list("requiresTools"),
        requires_env: string_list("requiresEnv"),
        fallback_for_tools: string_list("fallbackForTools"),
        agent_created: o.get("agent_created").and_then(serde_json::Value::as_bool).unwrap_or(false),
        created_at: o.get("created_at").and_then(serde_json::Value::as_u64),
        activity_ts: o.get("activity_ts").and_then(serde_json::Value::as_u64),
        usage_count: o.get("usage_count").and_then(serde_json::Value::as_u64).unwrap_or(0),
        archived: o.get("archived").and_then(serde_json::Value::as_bool).unwrap_or(false),
    })
}

/// Normalize an invocation handle (mirrors the frontend `normalizeHandle`):
/// trim → lowercase → whitespace runs become `-` → strip everything except
/// `[a-z0-9-]` (underscores are removed) → collapse dashes → trim leading
/// dashes.
pub fn normalize_handle(input: &str) -> String {
    let trimmed = input.trim().to_ascii_lowercase();
    let mut out = String::new();
    let mut last_was_space = false;
    for c in trimmed.chars() {
        if c.is_whitespace() {
            if !last_was_space && !out.is_empty() {
                out.push('-');
                last_was_space = true;
            }
        } else if c.is_ascii_alphanumeric() || c == '-' {
            out.push(c);
            last_was_space = false;
        } else {
            // Non-allowed char (underscore, symbol): drop it, mark a dash seam.
            last_was_space = true;
        }
    }
    let mut out = out.trim_matches('-').to_string();
    while out.contains("--") {
        out = out.replace("--", "-");
    }
    out
}

const HANDLE_RE_LEN: usize = 63;

/// Validate a skill create payload. Returns an error string or `None` when
/// valid. Mirrors `validateSkillFields`.
pub fn validate_skill_fields(input: &SkillCreateInput) -> Option<String> {
    if input.name.trim().is_empty() {
        return Some("skill name cannot be empty".into());
    }
    if input.prompt.trim().is_empty() {
        return Some("skill prompt cannot be empty".into());
    }
    if let Some(h) = input.handle.as_deref().map(str::trim).filter(|h| !h.is_empty()) {
        let ok = !h.is_empty()
            && h.len() <= HANDLE_RE_LEN
            && h.chars().next().is_some_and(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
            && h.chars().last().is_some_and(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
            && h
                .chars()
                .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_' || c == '-');
        if !ok {
            return Some(format!(
                "invalid handle '{h}': must match /^[a-z0-9][a-z0-9_-]{{0,62}}[a-z0-9]$/"
            ));
        }
    }
    if let Some(list) = &input.tool_allowlist {
        if list.iter().any(|t| t.trim().is_empty()) {
            return Some("toolAllowlist must contain only non-empty strings".into());
        }
    }
    None
}

#[derive(Debug, Default, Clone)]
pub struct SkillCreateInput {
    pub name: String,
    pub description: String,
    pub prompt: String,
    pub handle: Option<String>,
    pub tool_allowlist: Option<Vec<String>>,
}

/// Sanitize a skill directory/file name to `[a-zA-Z0-9._-]` (mirrors the
/// frontend `safeName`).
pub fn safe_skill_name(name: &str) -> String {
    name.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

/// Reject skill names that would escape the skills directory or collide with
/// hidden/system entries. Checks the raw name for path separators / traversal
/// before sanitization (a `../x` or `a/b` must never reach the fs layer).
pub fn skill_name_is_safe(name: &str) -> bool {
    if name.is_empty()
        || name == "."
        || name == ".."
        || name.starts_with('.')
        || name.contains('/')
        || name.contains('\\')
    {
        return false;
    }
    !safe_skill_name(name).is_empty()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_valid_skill_json() {
        let raw = r#"{"name":"fix-ts","description":"Fix TS errors","prompt":"Run pnpm check-types and fix","handle":"fixTs","toolAllowlist":["bash_run"],"agent_created":true,"usage_count":3}"#;
        let s = parse_skill_json(raw).unwrap();
        assert_eq!(s.name, "fix-ts");
        assert_eq!(s.handle.as_deref(), Some("fixts"));
        assert_eq!(s.tool_allowlist.as_deref(), Some(["bash_run".to_string()].as_slice()));
        assert!(s.agent_created);
        assert_eq!(s.usage_count, 3);
    }

    #[test]
    fn parse_rejects_missing_name_or_prompt() {
        assert!(parse_skill_json(r#"{"description":"x"}"#).is_none());
        assert!(parse_skill_json(r#"{"name":"x"}"#).is_none());
        assert!(parse_skill_json(r#"{"name":"  ","prompt":"p"}"#).is_none());
        assert!(parse_skill_json(r#"{"name":"n","prompt":"  "}"#).is_none());
        assert!(parse_skill_json("not json").is_none());
    }

    #[test]
    fn parse_tolerates_missing_optional_fields() {
        let s = parse_skill_json(r#"{"name":"n","prompt":"p"}"#).unwrap();
        assert_eq!(s.description, "");
        assert!(s.handle.is_none());
        assert!(!s.archived);
        assert_eq!(s.usage_count, 0);
    }

    #[test]
    fn normalize_handle_lowercases_and_collapses() {
        assert_eq!(normalize_handle("Fix-TS Errors"), "fix-ts-errors");
        assert_eq!(normalize_handle("  deploy  app  "), "deploy-app");
        assert_eq!(normalize_handle("UPPER"), "upper");
        assert_eq!(normalize_handle("a__b"), "ab");
        assert_eq!(normalize_handle("a  b"), "a-b");
        assert_eq!(normalize_handle("-leading-"), "leading");
    }

    #[test]
    fn validate_skill_fields_checks_rules() {
        assert!(validate_skill_fields(&SkillCreateInput {
            name: "x".into(),
            prompt: "p".into(),
            ..Default::default()
        })
        .is_none());
        assert!(validate_skill_fields(&SkillCreateInput {
            name: "".into(),
            prompt: "p".into(),
            ..Default::default()
        })
        .is_some());
        assert!(validate_skill_fields(&SkillCreateInput {
            name: "x".into(),
            prompt: "".into(),
            ..Default::default()
        })
        .is_some());
        // Handle with uppercase is invalid.
        assert!(validate_skill_fields(&SkillCreateInput {
            name: "x".into(),
            prompt: "p".into(),
            handle: Some("Bad".into()),
            ..Default::default()
        })
        .is_some());
        // Empty allowlist entries rejected.
        assert!(validate_skill_fields(&SkillCreateInput {
            name: "x".into(),
            prompt: "p".into(),
            tool_allowlist: Some(vec!["ok".into(), "  ".into()]),
            ..Default::default()
        })
        .is_some());
    }

    #[test]
    fn safe_name_and_safety_guard() {
        assert_eq!(safe_skill_name("My Skill!"), "My_Skill_");
        assert!(skill_name_is_safe("my-skill"));
        assert!(skill_name_is_safe("a.b"));
        assert!(!skill_name_is_safe(".."));
        assert!(!skill_name_is_safe("."));
        assert!(!skill_name_is_safe(".hidden"));
        assert!(!skill_name_is_safe("a/b"));
        assert!(!skill_name_is_safe("a\\b"));
        assert!(!skill_name_is_safe(""));
    }

    #[test]
    fn parse_requires_and_fallback_fields() {
        let raw = r#"{"name":"deploy","prompt":"deploy it","requiresTools":["bash_run"],"requiresEnv":["AWS_ACCESS_KEY_ID"],"fallbackForTools":["docker"]}"#;
        let s = parse_skill_json(raw).unwrap();
        assert_eq!(s.requires_tools.as_deref(), Some(["bash_run".to_string()].as_slice()));
        assert_eq!(s.requires_env.as_deref(), Some(["AWS_ACCESS_KEY_ID".to_string()].as_slice()));
        assert_eq!(s.fallback_for_tools.as_deref(), Some(["docker".to_string()].as_slice()));
    }

    #[test]
    fn skill_state_all_requirements_met_is_active() {
        let s = parse_skill_json(r#"{"name":"d","prompt":"p","requiresTools":["bash_run"]}"#).unwrap();
        assert_eq!(
            skill_state(&s, &["bash_run".to_string()], &[]),
            SkillState::Active
        );
    }

    #[test]
    fn skill_state_missing_requirement_is_degraded() {
        let s = parse_skill_json(r#"{"name":"d","prompt":"p","requiresEnv":["FOO"]}"#).unwrap();
        assert_eq!(skill_state(&s, &[], &[]), SkillState::Degraded);
        let t = parse_skill_json(r#"{"name":"d","prompt":"p","requiresTools":["git_commit"]}"#).unwrap();
        assert_eq!(skill_state(&t, &["read_file".to_string()], &[]), SkillState::Degraded);
    }

    #[test]
    fn skill_state_fallback_tool_missing_is_unavailable() {
        let s = parse_skill_json(r#"{"name":"d","prompt":"p","fallbackForTools":["docker"]}"#).unwrap();
        assert_eq!(skill_state(&s, &["read_file".to_string()], &[]), SkillState::Unavailable);
    }

    #[test]
    fn skill_state_no_requirements_is_active() {
        let s = parse_skill_json(r#"{"name":"d","prompt":"p"}"#).unwrap();
        assert_eq!(skill_state(&s, &[], &[]), SkillState::Active);
    }

    #[test]
    fn cap_skill_body_truncates_at_budget() {
        let short = "short body";
        assert_eq!(cap_skill_body(short, 100), short);
        let long = "x".repeat(5000);
        let capped = cap_skill_body(&long, SKILL_MAX_CHARS);
        assert_eq!(
            capped.chars().count(),
            SKILL_MAX_CHARS + 1 + "[truncated]".chars().count()
        );
        assert!(capped.contains("[truncated]"));
    }
}
